---
title: "Stat 115 2019: Homework 2"
author: "(your name)"
date: "Due: February 24, 2019"
output: html_document
---

```{r libraries, message = FALSE}
library(affy)
library(affyPLM)
library(limma)
library(hgu133plus2.db)
library(sva)
library(ggplot2)
library(dplyr)
library(biobroom)
library(class)
library(e1071)
```

# Part I. Differential Expression

In this part, we will continue to analyze the microarray gene expression
dataset from HW1.

1. The RMA run in HW1 will estimate the expression index of the samples
in a matrix. Use LIMMA, find the differentially expressed genes between
siEZH2 and control in LNCaP cells, and repeat the same in ABL cells. Use
false discovery rate (FDR) 0.05 and fold change (FC) 1.3 as cutoff to
filter the final result list. How many genes are differentially
expressed at this cutoff? Count in terms of gene symbol (e.g. TP53)
instead of transcript like (e.g. NM_000546)?

```{r part1-load, cache = TRUE}
celFiles <- list.celfiles(path = "data", full.names=TRUE)
data.affy <- ReadAffy(filenames = celFiles)
data.rma <- rma(data.affy)
expr.rma <- exprs(data.rma)
```

```{r part1-limma}
# your code here
```

2. Draw a hierarchical clustering of the 12 samples. Does the clustering
suggest the presence of batch effect?
Hint: use functions dist, hclust, plot

```{r part1-hclust}
# your code here
```

3. Use ComBat (included in the `sva` package) to adjust for batch
effects and provide evidence that the batch effects are successfully
adjusted. Repeat the differential expression analysis using LIMMA, FDR
0.05 and FC 1.3. Are there significant genes reported?

```{r part1-combat}
# your code here
```

4. FOR GRADUATES: Run K-means clustering of differentially expressed
genes across all 12 samples. Experiment with different K (there may not
be a correct answer here so just explore and explain your reasoning).
Hint: function kmeans

```{r part1-kmeans}
# your code here
```

5. Run the four list of differential genes (up / down, LNCaP / ABL)
separately on DAVID  (http://david.abcc.ncifcrf.gov/, you might want to
read their Nat Prot tutorial) to see whether the genes in each list are
enriched in specific biological process, pathways, etc. What’s in common
and what’s the most significant difference in EZH2 regulated genes
between LNCaP and ABL?

6. FOR GRADUATES: Try Gene Set Enrichment analysis
(http://www.broadinstitute.org/gsea/index.jsp) on the siEZH2 experiments
in LNCaP and ABL separately. Do the two cell lines differ in the
enriched gene sets?

```{r part1-gsea}
# your code here
```

# Part II. Microarray Clustering and Classification

The sample data is in file "taylor2010_data.txt" included in this
homework. This dataset has expression profiles of 23,974 genes in 27
normal samples, 129 primary cancer samples, 18 metastasized cancer
samples, and 5 unknown samples. Assume the data has been normalized and
summarized to expression index. The skeleton R code is provided here.

```{r loadtaylor}
taylor <- as.matrix(read.csv("data/taylor2010_data.txt", sep="\t",row.names=1))
index_normal <- grepl("N.P", colnames(taylor))
index_primary <- grepl("P.P", colnames(taylor))
index_met <- grepl("M.P", colnames(taylor))
n_normal <- sum(index_normal);
n_primary = sum(index_primary);
n_met = sum(index_met);

# class label (design vector)
taylor_classes = c(rep(0,n_normal), rep(1,n_primary), rep(2,n_met));

# train (known type samples), and test (unknown type samples)
train <- taylor[,1:174];
test <- taylor[,175:179];

tumortype_all <- factor(c(taylor_classes, rep(3, 5)), levels = 0:3,
                        labels = c("Normal", "Primary", "Metastasized",
                                   "Unknown"))
tumortype_class <- factor(taylor_classes, levels = 0:2,
                          labels = c("Normal", "Primary",
                                     "Metastasized"))

train_samps <- 1:174
test_samps <- 175:179
```

1. For the 174 samples with known type (normal, primary, metastasized),
use LIMMA to find the differentially expressed genes with fold change
threshold 1.3, and adjusted p-value threshold 0.05.  How many
differentially expressed genes are there? Hint: the design vector
consists of type indicator for the 174 samples. For example, 0 for
normal, 1 for primary, and 2 for metastasized.

```{r part2-limma}
# your code here
```

2. Perform k-means clustering on all samples using the differentially
expressed genes. Do the samples cluster according to disease status?

```{r part2-kmeans}
# your code here
```

3. Draw PCA biplot on the samples with differentially expressed genes,
and use 4 different colors to distinguish the 4 types of samples
(normal, primary, metastasized and unknown). Do the samples from
different groups look separable?
Hint: use ggplot

```{r part2-pca-biplot}
# your code here
```

4. FOR GRADUATES: What percent of variation in the data is captured in
the first two principle components? How many principle components do we
need to capture 80% of the variation in the data?
RHint: use function prcomp.

```{r part2-variation}
# your code here
```

5. Based on the PCA biplot, can you classify the 5 unknown samples?  Put
the PCA biplot in your HW write-up, and indicate which unknown sample
should be classified into which known type (normal, primary,
metastasized). Do you have different confidence for each unknown sample?

Your text answer here.

6. FOR GRADUATES: Use PCA on all samples and all the genes (instead of
the differentially expressed genes) for sample classification. Compare
to your results in the previous question. Which PCA plot looks better
and why?

```{r part2-pca2}
# your code here
```

7. Run KNN (try K = 1, 3 and 5) on the differential genes and all the
samples, and predict the unknown samples based on the 174 labeled
samples. Hint: use library class and function knn.

```{r part2-knn}
# your code here
```

8. Run SVM (try a linear kernel) on the differential genes and all the
samples, and predict the unknown samples based on the 174 labeled
samples. Hint: use library e1071 and function svm.

```{r part2-svm}
# your code here
```

9. FOR GRADUATES: Implement a 3-fold cross validation on your SVM
classifier, based on the 174 samples with known labels. What is your
average (of 3) classification error rate on the training data?

```{r part2-cv}
# your code here
```

# Part III. High throughput sequencing read mapping

We will give you a simple example to test high throughput sequencing
alignment for RNA-seq data. Normally for paired-end sequencing data,
each sample will have two separate FASTQ files, with line-by-line
correspondence to the two reads from the same fragment. Read mapping
could take a long time, so we have created just two FASTQ files of one
RNA-seq sample with only 3M fragments (2 * 3M reads) for you to run STAR
instead of the full data. The files are located at
`/n/stat115/HW2_2019/`. The mapping will generate one single output
file. Make sure to use the right parameters for single-end (SE) vs
paired-end (PE) modes in BWA and STAR.

Please include the commands that you used to run BWA and STAR in your
answers.

1. Use BWA (Li & Durbin, Bioinformatics 2009) to map the reads to the
Hg38 version of the reference genome, available on Odyssey at
`/n/stat115/HW2_2019/bwa_hg38_index/hg38.fasta`. Use the PE alignment
mode and generate the output in SAM format. Use SAMTools on the output
to find out how many fragments are mappable and uniquely mappable.

```
bwa commands here
```

2. Use STAR (Dobin et al, Bioinformatics 2012) to map the reads to the
reference genome, available on Odyssey at
`/n/stat115/HW2_2019/STARIndex`. Use the paired-end alignment mode and
generate the output in SAM format. STAR should have a report.  How many
fragments are mappable and how many are uniquely mappable?

```
STAR commands here
```

3. If you are getting a different number of mappable fragments between
BWA and STAR on the same data, why?

4. For GRADUATES: Run STAR using SE alignment mode on the left read
file. Take a look at the SE SAM file vs the PE SAM file. Are you getting
the same number of aligned fragments using PE mode vs SE mode?

```
STAR commands here
```

# Rules for submitting the homework:

Please submit your solution directly on the canvas website. Please
provide both your code in this Rmd document and an html file for your
final write-up. Please pay attention to the clarity and cleanness of
your homework.

The teaching fellows will grade your homework and give the grades with
feedback through canvas within one week after the due date. Some of the
questions might not have a unique or optimal solution. TFs will grade
those according to your creativity and effort on exploration, especially
in the graduate-level questions.


